toolkit/dataloader_mixins.py

```diff
@@ -25,13 +25,107 @@
 from PIL.ImageOps import exif_transpose
 import albumentations as A
 
-from toolkit.train_tools import get_torch_dtype
 
 if TYPE_CHECKING:
     from toolkit.data_loader import AiToolkitDataset
     from toolkit.data_transfer_object.data_loader import FileItemDTO
     from toolkit.stable_diffusion_model import StableDiffusion
 
+
+class EdgeMapFileItemDTOMixin:
+    def __init__(self: 'FileItemDTO', *args, **kwargs):
+        if hasattr(super(), '__init__'):
+            super().__init__(*args, **kwargs)
+        self.has_edge_map = False
+        self.edge_map_tensor: Union[torch.Tensor, None] = None
+        self.edge_map_cache_dir = '_edge_map_cache'
+        self.edge_detection_method = 'canny'  # Default method
+        self.canny_threshold1 = 100
+        self.canny_threshold2 = 200
+        dataset_config: 'DatasetConfig' = kwargs.get('dataset_config', None)
+        
+        if dataset_config.edge_map_path is not None:
+            img_path = kwargs.get('path', None)
+            file_name = os.path.basename(img_path)
+            file_dir = os.path.dirname(img_path)
+            file_name_without_ext = os.path.splitext(file_name)[0]
+            self.edge_map_path = os.path.join(file_dir, self.edge_map_cache_dir, f"{file_name_without_ext}_edge.png")
+            self.has_edge_map = True
+
+    def load_edge_map(self: 'FileItemDTO'):
+        if not self.has_edge_map:
+            return
+
+        if not os.path.exists(self.edge_map_path):
+            # Generate and save the edge map if it doesn't exist
+            img = self.load_image(self.path)
+            edge_map = self.generate_edge_map(img)
+            os.makedirs(os.path.dirname(self.edge_map_path), exist_ok=True)
+            cv2.imwrite(self.edge_map_path, edge_map)
+        else:
+            # Load the existing edge map
+            edge_map = cv2.imread(self.edge_map_path, cv2.IMREAD_GRAYSCALE)
+
+        # Convert to tensor
+        transform = transforms.Compose([
+            transforms.ToTensor(),
+        ])
+        if self.aug_replay_spatial_transforms:
+            self.edge_map_tensor = self.augment_spatial_control(Image.fromarray(edge_map), transform=transform)
+        else:
+            self.edge_map_tensor = transform(edge_map)
+
+    def load_image(self: 'FileItemDTO', image_path: str) -> np.ndarray:
+        img = Image.open(image_path)
+        img = exif_transpose(img)
+        img = img.convert('RGB')
+        return np.array(img)
+
+    def generate_edge_map(self: 'FileItemDTO', image: np.ndarray) -> np.ndarray:
```

Comment:
1. why here and not pre-extract to all?
2. I would not write the edge map function myself since extracting edge maps can work with few methods, also threshold matters. it's better to use the default method theyt leverage for this specific controlnet because this is how it was trained

Reply:
it does pre extract, cache_edge_maps